pkg/secretless/config/v2/service.go

```diff
@@ -0,0 +1,48 @@
+package v2
+
+import validation "github.com/go-ozzo/ozzo-validation"
+
+// ConnectorConfig is a wrapper around byte slice
+// that allows the connector configuration
+// to be Marshalled to YAML.
+type ConnectorConfig []byte
+
+func (c ConnectorConfig) MarshalYAML() (interface{}, error) {
+	return string(c), nil
+}
+
+func (c ConnectorConfig) Bytes() []byte {
+	return c
+}
+
+// Service represents a the configuration of a Secretless proxy service. It
+// includes the service's protocol, the socket or address it listens on, the
+// location of its required credentials, and (optionally) any additional
+// protocol specific configuration.
+type Service struct {
+	Debug 			bool
+	Connector       string
+	ConnectorConfig ConnectorConfig
+	Credentials     []*Credential
+	ListenOn        string
+	Name            string
+}
+
+// HasCredential indicates whether a Service has the specified credential.
+func (s Service) HasCredential(credentialName string) bool {
+	for _, credential := range s.Credentials {
+		if credential.Name == credentialName {
+			return true
+		}
+	}
+	return false
+}
+
+// Validate verifies the completeness and correctness of the Service.
+func (s Service) Validate() error {
+	return validation.ValidateStruct(&s,
+		validation.Field(&s.Credentials, validation.Required),
+		validation.Field(&s.ListenOn, validation.Required),
+		validation.Field(&s.Name, validation.Required),
+	)
+}
```

Comment:
When I did "Find Usages..." for this in GoLand, I got no results.  Which doesn't really matter....

Here's something that does matter imo, and that I don't think we're following everywhere but we should be.

We shouldn't have `Validate()` methods on our domain value objects, which is what `Service` is.... ozzo's README can be damned.  You validate user input, not domain objects.  What you really want is separate objects for each, and when you construct a value object it should actually be impossible to construct it invalidly.  You construct an immutable value object (which is always correct) using validated input.  

However, this would be verbose in Go, and kind of against the grain of the language.  But we can still act as if it's true and get the benefits of it, even if it would be possible to "go around" it.  We can still keep the `Validate()` function as a separate function (rather than a method on `Service`) and in our "border" code where we dealing with user input via config files, we call `Validate()` on our maybe invalid struct.  But after that happens, we have a full-fledged value object which we pass down to our application, which can treat it as immutable and always valid.  And so a `Validate()` method makes no sense.

Reply:
I think I understand what you're getting at here. It may well be possible to go around which is problematic.

Are you able to provide an example snippet showing how you'd resolve the concerns ?

Reply:
So because there was no usages of `Validate()` that I could find, there was no code to show.  But the only difference this would mean, if there was such code, is that you'd `Validate(yourService)` (free floating func defined in correct package) vs `yourService.Validate()` in the code that does input validation.  Now all the application code that uses `yourService` doesn't see this random `Validate()` method hanging around.  Since, again, that code is only applicable at the border when validation is done.